source/System.cpp

```diff
@@ -362,14 +354,43 @@ void System::UpdateNeighbors(const Set<System> &systems)
 
 
 
+// Once the star map is fully loaded, figure out which stars are "neighbors"
+// of this one, i.e. close enough to see or to reach via jump drive.
+void System::UpdateNeighbors(const Set<System> &systems, const double distance)
+{
+	set<const System *> &neighborSet = neighbors[distance];
+	
+	// Every star system that is linked to this one is automatically a neighbor,
+	// even if it is farther away than the maximum distance.
+	for(const System *system : links)
+		if(!(system->Position().Distance(position) <= distance))
+			neighborSet.insert(system);
+	
+	// Any other star system that is within the neighbor distance is also a
+	// neighbor. This will include any nearby linked systems.
+	for(const auto &it : systems)
+	{
+		// Skip systems that have no name.
+		if(it.first.empty() || it.second.Name().empty())
+			continue;
+
+		if(&it.second != this && it.second.Position().Distance(position) <= distance)
+			neighborSet.insert(&it.second);
+	}
+}
+
+
+
 // Modify a system's links.
 void System::Link(System *other)
 {
 	links.insert(other);
 	other->links.insert(this);
 	
-	neighbors.insert(other);
-	other->neighbors.insert(this);
+	for(auto &neighborSet : neighbors)
```

Comment:
>When there is only one set of neighbors, then it looks okay to add the hyperspace links and the jump links into one set, but when we have neighbor-sets per distance, then it looks a bit unnatural to add (and remove) hyperspace links to all the possible neighbor sets.

Given that jump drives use neighbor sets (they don’t actually ever use the links set), it’s important to update every set of neighbors, as the set of neighbors used depends on the jump drive.

Reply:
The jump drives can use the link sets without the hyperspace links being visible in them.

The code that does care is the code that determines if other systems are visible from some system. The loops there currently only use the neighbors and will need to be changed to use neighbors and links if the links are not contained in neighbors.

Reply:
I think that simply putting all linked systems within the neighbors as has been the case is simpler than updating every other instance in the code where jump drives are used to check both neighbors and links.

Reply:
Okay, the code by itself is not wrong and the neighbors code indeed already worked this way before this PR, so I can understand that you just want to keep it as-is for this PR.

Reply:
While we're here though, I think that updating neighbors in Link and Unlink is actually 100% unnecessary.
1. PlayerInfo::AddChange is where events are fed.
2. That function passes the changes to GameData::Change.
3. If an event changes links, then GameData calls these Link and Unlock function.
4. Returning to PlayerInfo::AddChange, if an event altered a system or changes links, then it calls GameData::UpdateSystems, which calls System::UpdateSystem.
5. UpdateSystem clears out all neighbors and then recalculates them.
Therefore, Link and Unlink don't need to update neighbors, since it's eventually done anyways.